logic/path/src/selectors/legacy.rs

```diff
@@ -75,73 +107,134 @@ impl<CW> DfsPathSelector<CW>
+            // We cannot use destination as last intermediate hop as
+            // this would be a loopback which does not give any privacy
+            return false;
+        }
+
+        // Check node reliability, new nodes are considered reliable
+        // unless the opposite has been observed
+        if channel.quality.unwrap_or(1.0f64) < self.quality_threshold {
+            // Only use nodes that have shown to be somewhat reliable
+            return false;
+        }
+
+        if current_path.contains(&channel.channel.destination) {
+            // At the moment, we do not allow circles because they
+            // do not give additional privacy
+            return false;
+        }
+
+        if U256::zero().eq(&channel.channel.balance.amount()) {
+            // We cannot use channels with zero stake
+            return false;
+        }
+
+        true
     }
 }
 
 impl<CW> PathSelector<CW> for DfsPathSelector<CW>
 where
     CW: EdgeWeighting<U256>,
 {
+    /// Attempts to find a path with at least `min_hops` hops and at most `max_hops` hops
+    /// that goes from `source` to `destination`. There does not need to be a
+    /// a payment channel to to `destination`, so the path only includes intermediate hops.
+    ///
+    /// Implements a randomized best-first search through the path space. The graph
+    /// traversal is bounded by `self.max_iterations` to prevent from long-running path
+    /// selection runs.
     fn select_path(
         &self,
         graph: &ChannelGraph,
         source: Address,
         destination: Address,
+        min_hops: usize,
         max_hops: usize,
     ) -> Result<ChannelPath> {
         if !(1..=INTERMEDIATE_HOPS).contains(&max_hops) {
             return Err(GeneralError::InvalidInput.into());
         }
 
         let mut queue = BinaryHeap::new();
-        let mut dead_ends = HashSet::new();
 
-        graph
-            .open_channels_from(source)
-            .filter_map(|channel| {
-                let w = channel.weight();
-                self.filter_channel(w, &destination, &[], &dead_ends)
-                    .then(|| WeightedChannelPath(vec![w.channel.destination], CW::calculate_weight(&w.channel)))
-            })
-            .for_each(|wcp| queue.push(wcp));
+        queue.extend(graph.open_channels_from(source).filter_map(|channel| {
+            let w = channel.weight();
+            self.filter_channel(w, &source, &destination, &[])
+                .then(|| WeightedChannelPath {
+                    path: vec![w.channel.destination],
+                    weight: CW::calculate_weight(&w.channel),
+                    fully_explored: false,
+                })
+        }));
 
         let mut iters = 0;
-        while !queue.is_empty() && iters < self.max_iterations {
-            let current_path = queue.peek().unwrap();
-            let current_path_len = current_path.0.len();
+        while let Some(mut current_path) = queue.pop() {
```

Comment:
This is a for cycle over a `..=self.max_iterations`

Reply:
added a possible recovery